packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl

```diff
@@ -12,9 +12,9 @@ systemd:
             Environment="ETCD_NAME=${etcd_name}"
             Environment="ETCD_ADVERTISE_CLIENT_URLS=https://${etcd_domain}:2379"
             Environment="ETCD_INITIAL_ADVERTISE_PEER_URLS=https://${etcd_domain}:2380"
-            Environment="ETCD_LISTEN_CLIENT_URLS=https://0.0.0.0:2379"
-            Environment="ETCD_LISTEN_PEER_URLS=https://0.0.0.0:2380"
-            Environment="ETCD_LISTEN_METRICS_URLS=http://0.0.0.0:2381"
+            Environment="ETCD_LISTEN_CLIENT_URLS=https://REPLACE_WITH_PRIVATE_IP:2379"
+            Environment="ETCD_LISTEN_PEER_URLS=https://REPLACE_WITH_PRIVATE_IP:2380"
+            Environment="ETCD_LISTEN_METRICS_URLS=http://REPLACE_WITH_PRIVATE_IP:2381"
```

Comment:
What? Why not use `COREOS_PACKET_IPV4_PRIVATE_0` var? Like [here](https://github.com/kinvolk/lokomotive-kubernetes/blob/ae523cf4df053d6b3cde113b9f947a5d2d10f052/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl#L96)

You need [this line](https://github.com/kinvolk/lokomotive-kubernetes/blob/ae523cf4df053d6b3cde113b9f947a5d2d10f052/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl#L70) but... isn't there any reason that would not work?

Reply:
This did not work, we might have to move the parameters from env to cmd line flags. That means also provide `ExecStart`.

I also would like to know why etcd service is shipped without any cmd line flags?

Reply:
Not sure I follow, why this didn't work?

And I guess the env vars are the params, right?

Reply:
With this in `/run/metadata`:

```
COREOS_PACKET_IPV4="1.2.3.4"
```

this doesn't work:

```
[Unit]
Description=Test

[Service]
EnvironmentFile=/run/metadata
Environment="OMG_LISTEN=${COREOS_PACKET_IPV4}"
ExecStart=/bin/bash -c 'env'
```
env shows `OMG_LISTEN=${COREOS_PACKET_IPV4}`

however this works:

```
[Unit]
Description=Test

[Service]
EnvironmentFile=/run/metadata
Environment="OMG_LISTEN=${COREOS_PACKET_IPV4}"
ExecStart=/bin/bash -c 'OMG_LISTEN=${COREOS_PACKET_IPV4} env'
```

env shows `OMG_LISTEN=1.2.3.4`

So it seems systemd only does substitutions inside `ExecStart`.

Reply:
Ohhhh, I understand now. Doesn't seem very nice either :-(

Reply:
Ok so falling back on doing `curl` then?

Reply:
Not sure, my gut feeling is that we can use the flatcar metadata service in a cleaner way than the curl thing. And is more reliable, it retries and everything automatically.

For example, looking quickly at etcd manpage it seems we can also use the parameter `--listen-client-urls` to the command, instead of the env variable, and there we can use this expansion IIUC.

Probably is cleaner to move this from env vars to arguments and use the env var there.

Also, other alternatives might be possible and cleaner than curl too. Like having a systemd bash script that just reads the Flatcar metadata file (using the environment file) and populates the env as this unit is using it. Something like: `printf "ETCD_LISTEN_CLIENT_URLS=\"https://${COREOS_PRIVATE_IP_0}...\"`.

But I think the first option, just moving the env vars to command line arguments (i.e instead of env var `ETCD_LISTEN_CLIENT_URLS` use param `--listen-client-urls` and the same for others, hopefully there is param for all :crossed_fingers: ) seems way cleaner, IMHO.

What do you think?

Reply:
btw, sysmted documentation on why this is not expanded is here: https://www.freedesktop.org/software/systemd/man/systemd.exec.html

Reply:
Oh, or maybe adding args to the cmdline is not easy because we are using etcd-wrapper, right? I guess I lean towards creating the env file for this systemd unit, then.

But as you think it is better

Reply:
@rata how about now?

Reply:
@surajssd IMHO, it seems a little bit simpler. What do you think?

Not sure if it can be simpler, though. Like an inline systemd unit that has the flatcar metadata environment (so the COREOS_* vars are there) and executes something like:

```
printf "ETCD_LISTEN_CLIENT_URLS=https://%s:2379\n"  $COREOS_PACKET_IPV4_PRIVATE_0
printf "ETCD_LISTEN_PEER_URLS=https://%s:2380\n"    $COREOS_PACKET_IPV4_PRIVATE_0
printf "ETCD_LISTEN_METRICS_URLS=https://%s:2381\n" $COREOS_PACKET_IPV4_PRIVATE_0
```

redirecting the output to a file, and that file is loaded in the environment here? Maybe this intermediate file makes this option not so nice nor readable?

Reply:
done @rata

Reply:
Thanks!

But please, not do everything, share your opinion too! It's okay to disagree! =)